### DISCUSSION TITLE
Installing verity image takes long time when CODESYS is running.

### DISCUSSION THREAD

**Discussion by kjeldflarup** [rauc/rauc#1409] [Q&A]
We have 300 MB large RAUC bundles, which normally installs in 15 seconds from a local file. But recently we observed that it takes over a minute. Someone found out that, if CODESYS runtime was stopped it installed faster.

CODESYS is a PLC programming system, where it is possible to load applications, but I did not even have to load an application, just let CODESYS stand idle. 

I tried to install an old version of the RAUC file, before we started to use verity, and it installed in 7 seconds.

I have tried with both lttng and strace to find out what happens. lttng just showed that it ended up in a long lastning kworker session.

Strace gave more information as I got a number of these. It looks like rauc is polling for something
```
[pid 32775] 14:39:59 <... ppoll resumed>) = 1 ([{fd=3, revents=POLLIN}]) <8.493725>
[pid 32777] 14:40:01 <... ppoll resumed>) = 1 ([{fd=5, revents=POLLIN}]) <1.284844>
[pid 32775] 14:40:01 <... ppoll resumed>) = 1 ([{fd=3, revents=POLLIN}]) <1.285930>
[pid 32777] 14:40:01 ppoll([{fd=5, events=POLLIN}, {fd=6, events=POLLIN}], 2, NULL, NULL, 0) = 1 ([{fd=5, revents=POLLIN}]) <7.927204>
```
I cannot really se from the strace what the file descriptors are. Actually there were no fd=6
[rauc.strace.txt](https://github.com/rauc/rauc/files/15236087/rauc.strace.txt)

The poll it self is not found in the Rauc codebase, but must be used from https://gitlab.gnome.org/GNOME/glib/ which is used for the streaming (which I don't do)

Anybody seen a similar behaviour?
Somehow CODESYS and Rauc must be competing on a shared resource.

**Comment by jluebbe** [comment]
I assume that you're not using streaming via HTTP, right?

During installation of a bundle, multiple process may be involved and read from the mounted bundle, so it could be that a tar process is waiting for data and you're not seeing that when strace-ing rauc without `-f`. Also, it seems that your strace log is from the rauc CLI, not the service, so it mostly waits for progress messages from the service and doesn't show the interesting parts.

You could try `rauc mount` to exclude RAUC's installation code and see if there are performance differences when reading/copying manually from the bundle with or without CODESYS.

You could try `blktrace` to see if there are read or write IOs that are delayed significantly.

How is your kernel configured with regards to `CONFIG_SQUASHFS_*`?

You mention that CODESYS is idle in this case, so it's not consuming much CPU time? Does it consume a lot of memory? Perhaps take a look at differences in `/proc/meminfo`.

You could check what CODESYS is doing with `strace -f -p <pid>`. Perhaps it's calling `sync`/`syncfs` or some other syscall which has system-wide effects?
